lib/db/schema/patch-032-033.sql

```diff
@@ -0,0 +1,96 @@
+CREATE TABLE IF NOT EXISTS securityEventNames (
+  id INT AUTO_INCREMENT PRIMARY KEY,
+  name VARCHAR(255) NOT NULL,
+  UNIQUE INDEX securityEventNamesUnique(name)
+) ENGINE=InnoDB;
+
+CREATE TABLE IF NOT EXISTS securityEvents (
+  uid BINARY(16) NOT NULL,
+  nameId INT NOT NULL,
+  FOREIGN KEY (nameId) REFERENCES securityEventNames(id) ON DELETE CASCADE,
+  tokenId BINARY(32),
+  verified BOOLEAN,
+  ipAddrHmac BINARY(32) NOT NULL,
+  createdAt BIGINT SIGNED NOT NULL,
+  INDEX securityEvents_uid_ipAddrHmac (uid, ipAddrHmac)
+) ENGINE=InnoDB;
+
+INSERT INTO
+    securityEventNames (name)
+VALUES
+    ("account.create"),
+    ("account.login"),
+    ("account.reset");
+
+CREATE PROCEDURE `createSecurityEvent_1` (
+    IN inUid BINARY(16),
+    IN inToken BINARY(32),
+    IN inName INT,
+    IN inIpAddr BINARY(32),
+    IN inCreatedAt BIGINT SIGNED
+)
+BEGIN
+    INSERT INTO securityEvents(
+        uid,
+        tokenId,
+        verified,
+        nameId,
+        ipAddrHmac,
+        createdAt
+    )
+    VALUES(
+        inUid,
+        inToken,
+        (
+          SELECT COUNT(*)
+          FROM unverifiedTokens u
+          WHERE u.uid = inUid AND u.tokenId = inToken
+        ) = 0,
+        inName,
+        inIpAddr,
+        inCreatedAt
+    );
+END;
+
+CREATE PROCEDURE `fetchSecurityEvents_1` (
+    IN inUid BINARY(16),
+    IN inIpAddr BINARY(32)
+)
+BEGIN
+    SELECT
+        n.name,
+        e.verified,
+        e.createdAt
+    FROM
+        securityEvents e
+    LEFT JOIN securityEventNames n
+        ON e.nameId = n.id
+    WHERE
+        e.uid = inUid
+    AND
+        e.ipAddrHmac = inIpAddr;
+END;
+
+CREATE PROCEDURE `verifyToken_2` (
+  IN `tokenVerificationIdArg` BINARY(16),
+  IN `uidArg` BINARY(16)
+)
+BEGIN
+  UPDATE securityEvents
+  SET verified = true
+  WHERE (
+    SELECT tokenId
+    FROM unverifiedTokens
+    WHERE tokenVerificationId = tokenVerificationIdArg
+    AND uid = uidArg
+    LIMIT 1
+  )
```

Comment:
My SQL-fu could be failing me, but I don't really understand how this subquery works.  It appears to be selecting the`tokenId` that corresponds to the given `tokenVerificationIdArg`, which makes sense, but I don't see how it then compares that to the tokenIds in the rows of the securityEvents table so that we only updates the ones we want.

It may be clearer to write `WHERE tokenId = ( SELECT ... )` so that it's obvious that we want to filter on the matching rows of the `securityEvents` table.

This query is also slightly unusual in that there's no index covering `securityEvents.tokenId`, so it will have to scan through all the security events matching the given `userId` to find the rows to update.  IIRC this is on purpose, to avoid the overhead of maintaining another index.  But our future selves will probably appreciate a comment to that effect in the source here.

If you haven't done so, it would also be worth doing an `EXPLAIN` on this query to sanity-check that it can use the (uid, ipAddrHmac) index to walk just the security events for the given user.